can we add hexagon as shape for node in network

I am using the **Network** layout and I wanted to show node in **hexagon shape**, how can I do this, can we use SVG script in vis js

$ gulp lint
module.js:328
    throw err;
    ^

Error: Cannot find module 'gulp-concat'
    at Function.Module._resolveFilename (module.js:326:15)
    at Function.Module._load (module.js:277:25)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/home/rgupta/github/vis/gulpfile.js:5:14)
    at Module._compile (module.js:410:26)
    at Object.Module._extensions..js (module.js:417:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)

gulp lint throws an error

OK, following you here. Did the following:

- got the `develop` branch
- `> npm install`
- `> gulp lint`

I see the following `errors`:

```
[08:12:14] Using gulpfile ~/projects/github/rg2609/vis/gulpfile.js
[08:12:14] Starting 'lint'...
[08:12:17] 
/home/wim/projects/github/rg2609/vis/lib/network/modules/components/nodes/shapes/Hexagon.js
   5:1   error  Missing JSDoc comment  require-jsdoc
   6:14  error  Missing JSDoc comment  require-jsdoc
  10:7   error  Missing JSDoc comment  require-jsdoc
  14:19  error  Missing JSDoc comment  require-jsdoc

✖ 4 problems (4 errors, 0 warnings)

[08:12:17] 'lint' errored after 3.38 s
[08:12:17] ESLintError in plugin 'gulp-eslint'
Message:
    Failed with 4 errors
```
Is this the same that you get?

This means that you'll need to add some commenting to your added classes and methods. It's not a code error. You can skip this for now.

Now, try compiling a `vis.js` with your additions:

```
> gulp
```

On my machine, it builds perfectly. If that's the same for you, try out the examples as I described above:



    run a node example, e.g. examples/nodes/basicUsage.html
    change one of the node shapes to hexagon, e.g.:

    {id: 1, label: 'Node 1', shape:'hexagon'},

...and refresh.

--- a/docs/network/nodes.html
+++ b/docs/network/nodes.html
@@ -834,7 +834,7 @@ network.setOptions(options);
                 <code>ellipse</code>, <code>circle</code>, <code>database</code>, <code>box</code>, <code>text</code>.
                 Always scalable are:
                 <code>image</code>, <code>circularImage</code>, <code>diamond</code>, <code>dot</code>,
-                <code>star</code>, <code>triangle</code>, <code>triangleDown</code>, <code>square</code> and
+                <code>star</code>, <code>triangle</code>, <code>triangleDown</code>, <code>hexagon</code>, <code>square</code> and
                 <code>icon</code>. Keep in mind that when using scaling, the <code>size</code> option is neglected.
             </td>
         </tr>
@@ -975,7 +975,7 @@ mySize = minSize + diff * scale;
                 <code>ellipse</code>, <code>circle</code>, <code>database</code>, <code>box</code>, <code>text</code>.
                 The ones with the label outside of it are: <code>image</code>, <code>circularImage</code>,
                 <code>diamond</code>, <code>dot</code>, <code>star</code>, <code>triangle</code>,
-                <code>triangleDown</code>, <code>square</code> and <code>icon</code>.
+                <code>triangleDown</code>, <code>hexagon</code>, <code>square</code> and <code>icon</code>.
             </td>
         </tr>
         <tr class='toggle collapsible' onclick="toggleTable('optionTable','shapeProperties', this);">
@@ -1031,7 +1031,7 @@ mySize = minSize + diff * scale;
             <td><code>25</code></td>
             <td>The size is used to determine the size of node shapes that do not have the label inside of them. These
                 shapes are: <code>image</code>, <code>circularImage</code>, <code>diamond</code>, <code>dot</code>,
-                <code>star</code>, <code>triangle</code>, <code>triangleDown</code>, <code>square</code> and
+                <code>star</code>, <code>triangle</code>, <code>triangleDown</code>, <code>hexagon</code>, <code>square</code> and
                 <code>icon</code></td>
         </tr>
         <tr>

--- a/lib/network/modules/components/Node.js
+++ b/lib/network/modules/components/Node.js
@@ -12,6 +12,7 @@ var Ellipse = require('./nodes/shapes/Ellipse').default;
 var Icon = require('./nodes/shapes/Icon').default;
 var Image = require('./nodes/shapes/Image').default;
 var Square = require('./nodes/shapes/Square').default;
+var Hexagon = require('./nodes/shapes/Hexagon').default;
 var Star = require('./nodes/shapes/Star').default;
 var Text = require('./nodes/shapes/Text').default;
 var Triangle = require('./nodes/shapes/Triangle').default;

--- a/lib/network/modules/components/Node.js
+++ b/lib/network/modules/components/Node.js
@@ -369,6 +369,9 @@ class Node {
         case 'square':
           this.shape = new Square(this.options, this.body, this.labelModule);
           break;
+        case 'hexagon':
+          this.shape = new Hexagon(this.options, this.body, this.labelModule);
+          break;
         case 'star':
           this.shape = new Star(this.options, this.body, this.labelModule);
           break;

--- a/lib/network/modules/components/nodes/shapes/Hexagon.js
+++ b/lib/network/modules/components/nodes/shapes/Hexagon.js
@@ -2,15 +2,40 @@
 
 import ShapeBase from '../util/ShapeBase'
 
+/**
+ * A Hexagon Node/Cluster shape.
+ *
+ * @extends ShapeBase
+ */
 class Hexagon extends ShapeBase {
+  /**
+   * @param {Object} options
+   * @param {Object} body
+   * @param {Label} labelModule
+   */
   constructor(options, body, labelModule) {
     super(options, body, labelModule)
   }
 
+  /**
+   *
+   * @param {CanvasRenderingContext2D} ctx
+   * @param {number} x width
+   * @param {number} y height
+   * @param {boolean} selected
+   * @param {boolean} hover
+   * @param {{toArrow: boolean, toArrowScale: (allOptions.edges.arrows.to.scaleFactor|{number}|allOptions.edges.arrows.middle.scaleFactor|allOptions.edges.arrows.from.scaleFactor|Array|number), toArrowType: *, middleArrow: boolean, middleArrowScale: (number|allOptions.edges.arrows.middle.scaleFactor|{number}|Array), middleArrowType: (allOptions.edges.arrows.middle.type|{string}|string|*), fromArrow: boolean, fromArrowScale: (allOptions.edges.arrows.to.scaleFactor|{number}|allOptions.edges.arrows.middle.scaleFactor|allOptions.edges.arrows.from.scaleFactor|Array|number), fromArrowType: *, arrowStrikethrough: (*|boolean|allOptions.edges.arrowStrikethrough|{boolean}), color: undefined, inheritsColor: (string|string|string|allOptions.edges.color.inherit|{string, boolean}|Array|*), opacity: *, hidden: *, length: *, shadow: *, shadowColor: *, shadowSize: *, shadowX: *, shadowY: *, dashes: (*|boolean|Array|allOptions.edges.dashes|{boolean, array}), width: *}} values
+   */
   draw(ctx, x, y, selected, hover, values) {
     this._drawShape(ctx, 'hexagon', 4, x, y, selected, hover, values);
   }
 
+  /**
+   *
+   * @param {CanvasRenderingContext2D} ctx
+   * @param {number} angle
+   * @returns {number}
+   */
   distanceToBorder(ctx, angle) {
     return this._distanceToBorder(ctx,angle);
   }

--- a/lib/network/modules/components/nodes/shapes/Hexagon.js
+++ b/lib/network/modules/components/nodes/shapes/Hexagon.js
@@ -1,0 +1,19 @@
+'use strict';
+
+import ShapeBase from '../util/ShapeBase'
+
+class Hexagon extends ShapeBase {
+  constructor(options, body, labelModule) {
+    super(options, body, labelModule)
+  }
+
+  draw(ctx, x, y, selected, hover, values) {
+    this._drawShape(ctx, 'hexagon', 4, x, y, selected, hover, values);
+  }
+
+  distanceToBorder(ctx, angle) {
+    return this._distanceToBorder(ctx,angle);
+  }
+}
+
+export default Hexagon;

--- a/lib/network/options.js
+++ b/lib/network/options.js
@@ -319,7 +319,7 @@ let allOptions = {
       y: { number },
       __type__: { object, boolean: bool }
     },
-    shape: { string: ['ellipse', 'circle', 'database', 'box', 'text', 'image', 'circularImage', 'diamond', 'dot', 'star', 'triangle', 'triangleDown', 'square', 'icon'] },
+    shape: { string: ['ellipse', 'circle', 'database', 'box', 'text', 'image', 'circularImage', 'diamond', 'dot', 'star', 'triangle', 'triangleDown', 'square', 'icon', 'hexagon'] },
     shapeProperties: {
       borderDashes: { boolean: bool, array },
       borderRadius: { number },
@@ -466,7 +466,7 @@ let configureOptions = {
       x: [5, -30, 30, 1],
       y: [5, -30, 30, 1]
     },
-    shape: ['ellipse', 'box', 'circle', 'database', 'diamond', 'dot', 'square', 'star', 'text', 'triangle', 'triangleDown'],
+    shape: ['ellipse', 'box', 'circle', 'database', 'diamond', 'dot', 'square', 'star', 'text', 'triangle', 'triangleDown','hexagon'],
     shapeProperties: {
       borderDashes: false,
       borderRadius: [6, 0, 20, 1],

--- a/lib/network/shapes.js
+++ b/lib/network/shapes.js
@@ -332,7 +332,7 @@ if (typeof CanvasRenderingContext2D !== 'undefined') {
     var a = (Math.PI*2)/sides;
     this.moveTo(x+r,y);
     for (var i = 1; i < sides; i++) {
-      this.lineTo(r*Math.cos(a*i),r*Math.sin(a*i));
+      this.lineTo(x+r*Math.cos(a*i),y+r*Math.sin(a*i));
     }
       this.closePath();
   };

--- a/lib/network/shapes.js
+++ b/lib/network/shapes.js
@@ -328,13 +328,13 @@ if (typeof CanvasRenderingContext2D !== 'undefined') {
    */
   CanvasRenderingContext2D.prototype.hexagon = function (x, y, r) {
     this.beginPath();
-    var sides = 6
+    var sides = 6;
     var a = (Math.PI*2)/sides;
     this.moveTo(x+r,y);
     for (var i = 1; i < sides; i++) {
       this.lineTo(x+r*Math.cos(a*i),y+r*Math.sin(a*i));
     }
-      this.closePath();
+    this.closePath();
   };
 
 

--- a/lib/network/shapes.js
+++ b/lib/network/shapes.js
@@ -320,5 +320,22 @@ if (typeof CanvasRenderingContext2D !== 'undefined') {
     }
   };
 
+  /**
+   * Draw a Hexagon shape with 6 sides
+   * @param {Number} x horizontal center
+   * @param {Number} y vertical center
+   * @param {Number} r   radius
+   */
+  CanvasRenderingContext2D.prototype.hexagon = function (x, y, r) {
+    this.beginPath();
+    var a = (Math.PI*2)/6;
+    this.translate(x,y);
+    this.moveTo(radius,0);
+    for (var i = 1; i < sides; i++) {
+      this.lineTo(radius*Math.cos(a*i),radius*Math.sin(a*i));
+    }
+      this.closePath();
+  };
+
 
 }

--- a/lib/network/shapes.js
+++ b/lib/network/shapes.js
@@ -328,11 +328,12 @@ if (typeof CanvasRenderingContext2D !== 'undefined') {
    */
   CanvasRenderingContext2D.prototype.hexagon = function (x, y, r) {
     this.beginPath();
-    var a = (Math.PI*2)/6;
+    var sides = 6
+    var a = (Math.PI*2)/sides;
     this.translate(x,y);
-    this.moveTo(radius,0);
+    this.moveTo(r,0);
     for (var i = 1; i < sides; i++) {
-      this.lineTo(radius*Math.cos(a*i),radius*Math.sin(a*i));
+      this.lineTo(r*Math.cos(a*i),r*Math.sin(a*i));
     }
       this.closePath();
   };

--- a/lib/network/shapes.js
+++ b/lib/network/shapes.js
@@ -330,7 +330,6 @@ if (typeof CanvasRenderingContext2D !== 'undefined') {
     this.beginPath();
     var sides = 6
     var a = (Math.PI*2)/sides;
-    this.translate(x,y);
     this.moveTo(x+r,y);
     for (var i = 1; i < sides; i++) {
       this.lineTo(r*Math.cos(a*i),r*Math.sin(a*i));

--- a/lib/network/shapes.js
+++ b/lib/network/shapes.js
@@ -331,7 +331,7 @@ if (typeof CanvasRenderingContext2D !== 'undefined') {
     var sides = 6
     var a = (Math.PI*2)/sides;
     this.translate(x,y);
-    this.moveTo(r,0);
+    this.moveTo(x+r,y);
     for (var i = 1; i < sides; i++) {
       this.lineTo(r*Math.cos(a*i),r*Math.sin(a*i));
     }

--- a/test/Network.test.js
+++ b/test/Network.test.js
@@ -190,7 +190,15 @@ describe('Network', function () {
 
 
   after(function() {
-    this.jsdom_global();
+    try {
+      this.jsdom_global();
+    } catch(e) {
+      if (e.message() === 'window is undefined') {
+        console.warning("'" + e.message() + "' happened again");
+      } else {
+        throw e;
+      }
+    }
   });
 
 

